src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -867,7 +866,19 @@ evalApp γ _ctx e0 es _et
         -- This case is pretty sus as we shouldnt generate funcitons that do not
         -- have a type annotation
         Nothing -> do
-            traceShowM ("I'm failing on" :: String, f)
+          let etaArgsType = fmap snd $ drop nProvidedArgs $ eqArgs eq
+          -- Hack to recover general mono signature
+          if not $ isAnyTyVar etaArgsType then do
+          let etaVars = zipWith (\name ty -> ECst (EVar name) ty) etaNames etaArgsType
+          let fullBody = eApps e0 (es ++ etaVars)
+          let etaExpandedTerm = wrap fullBody (zip etaNames etaArgsType)
```

Comment:
Looks like defining an auxiliary function for eta expansion could eliminate some duplication.